src/coreclr/vm/reflectioninvocation.cpp

```diff
@@ -1166,8 +1095,61 @@ FCIMPL5(Object*, RuntimeMethodHandle::InvokeMethod,
 
     }
 
-Done:
-    ;
+    HELPER_METHOD_FRAME_END();
+
+    return OBJECTREFToObject(gc.retVal);
+}
+FCIMPLEND
+
+
+FCIMPL2(Object*, RuntimeMethodHandle::InvokeArrayCtor,
+    Span<OBJECTREF>* objs,
+    SignatureNative* pSigUNSAFE)
+{
+    FCALL_CONTRACT;
+
+    struct
+    {
+        SIGNATURENATIVEREF pSig;
+        OBJECTREF retVal;
+    } gc;
+
+    gc.pSig = (SIGNATURENATIVEREF)pSigUNSAFE;
+    gc.retVal = NULL;
+
+    TypeHandle th = gc.pSig->GetDeclaringType();
+    int argCnt = gc.pSig->NumFixedArgs();
+
+    HELPER_METHOD_FRAME_BEGIN_RET_PROTECT(gc);
+
+    // Validate the argCnt an the Rank. Also allow nested SZARRAY's.
+    _ASSERTE(argCnt == (int) th.GetRank() || argCnt == (int) th.GetRank() * 2 ||
+             th.GetInternalCorElementType() == ELEMENT_TYPE_SZARRAY);
+
+    // Validate all of the parameters.  These all typed as integers
+    int allocSize = 0;
+    if (!ClrSafeInt<int>::multiply(sizeof(INT32), argCnt, allocSize))
+        COMPlusThrow(kArgumentException, IDS_EE_SIGTOOCOMPLEX);
+
+    INT32* indexes = (INT32*) _alloca((size_t)allocSize);
+    ZeroMemory(indexes, allocSize);
+
+    for (DWORD i = 0; i < (DWORD)argCnt; i++)
+    {
+        if (!objs->GetAt(i))
+            COMPlusThrowArgumentException(W("parameters"), W("Arg_NullIndex"));
+
+        MethodTable* pMT = objs->GetAt(i)->GetMethodTable();
+        CorElementType oType = TypeHandle(pMT).GetVerifierCorElementType();
+
+        if (!InvokeUtil::IsPrimitiveType(oType) || !InvokeUtil::CanPrimitiveWiden(ELEMENT_TYPE_I4,oType))
+            COMPlusThrow(kArgumentException,W("Arg_PrimWiden"));
+
+        memcpy(&indexes[i], objs->GetAt(i)->UnBox(), pMT->GetNumInstanceFieldBytes());
```

Comment:
I know that this was here before, but it is wrong. The widening has to do proper sign-extension like what the regular invoke does (regular PInvoke does it using `InvokeUtil::CopyArg` method).

For example:
```csharp
using System.Reflection;

var ctor = typeof(int[,]).GetConstructor(new[] { typeof(int), typeof(int), typeof(int), typeof(int) });
var arr = (int[,])ctor.Invoke(new object[] { (sbyte)-1, (sbyte)2, (sbyte)-3, (sbyte)4 });
Console.WriteLine(arr.GetLowerBound(0));
Console.WriteLine(arr.GetLowerBound(1));
```
prints 255 253, but it should print -1 -3 instead.

Reply:
This is a bit clunky. It would appear the `InvokeUtil::CopyArg` and `InvokeUtil::CreatePrimitiveValue` both have bugs in them. They indiscriminately copy everything into the `ARG_SLOT` which for this case isn't the size of the destination. We are using a stack allocation and since `ARG_SLOT` is always 64-bits it corrupts the stack. I've fixed `InvokeUtil::CreatePrimitiveValue` to do the right thing.

Reply:
`ARG_SLOT` is what regular method invoke arguments get normalized into so I think it makes sense that `InvokeUtil::CopyArg` and `InvokeUtil::CreatePrimitiveValue` expect the destination to be `ARG_SLOT`.

Reply:
Right. I missed how this API was being used. I found an example of the correct way to use this and am reverting my changes to it.